bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryClientImpl.java

```diff
@@ -0,0 +1,492 @@
+                      }
+                      return true;
+                    });
+      }
+
+      return tablesStream.collect(Collectors.toList());
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  @SuppressWarnings("FormatStringAnnotation")
+  private Dataset updateDataset(Dataset dataset) {
+    try {
+      HttpResponse response =
+          client
+              .datasets()
+              .update(
+                  Preconditions.checkNotNull(dataset.getDatasetReference()).getProjectId(),
+                  Preconditions.checkNotNull(dataset.getDatasetReference().getDatasetId()),
+                  dataset)
+              .setRequestHeaders(new HttpHeaders().setIfMatch(dataset.getEtag()))
+              .executeUnparsed();
+      if (response.getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+        throw new NoSuchNamespaceException(response.getStatusMessage());
+      }
+
+      return convertExceptionIfUnsuccessful(response).parseAs(Dataset.class);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  /**
+   * Returns true when it is a BigQuery Metastore Iceberg table, defined by having the
+   * ExternalCatalogTableOptions object and a parameter of METADATA_LOCATION_PROP as part of its
+   * parameters map.
+   *
+   * @param table to check
+   */
+  private boolean isValidIcebergTable(Table table) {
+    return table.getExternalCatalogTableOptions() != null
+        && !table.getExternalCatalogTableOptions().isEmpty()
+        && table.getExternalCatalogTableOptions().getParameters() != null
+        && table
+            .getExternalCatalogTableOptions()
+            .getParameters()
+            .containsKey(BaseMetastoreTableOperations.METADATA_LOCATION_PROP)
+        && BaseMetastoreTableOperations.ICEBERG_TABLE_TYPE_VALUE.equalsIgnoreCase(
+            table
+                .getExternalCatalogTableOptions()
+                .getParameters()
+                .get(BaseMetastoreTableOperations.TABLE_TYPE_PROP));
+  }
+
+  private Table validateTable(Table table) {
+    if (!isValidIcebergTable(table)) {
+      throw new NoSuchIcebergTableException("This table is not a valid Iceberg table: %s", table);
+    }
+
+    return table;
+  }
+
+  /**
+   * Converts BigQuery generic API errors to Iceberg exceptions, *without* handling the
+   * resource-specific exceptions like NoSuchTableException, NoSuchNamespaceException, etc.
+   */
+  @SuppressWarnings("FormatStringAnnotation")
+  private HttpResponse convertExceptionIfUnsuccessful(HttpResponse response) throws IOException {
+    if (response.isSuccessStatusCode()) {
+      return response;
+    }
+
+    String errorMessage =
+        String.format(
+            "%s\n%s",
+            response.getStatusMessage(),
+            response.getContent() != null
+                ? new String(response.getContent().readAllBytes(), StandardCharsets.UTF_8)
+                : "");
```

Comment:
Nit: If there is no content, this will still add a newline. That seems unintentional.

Since there isn't much formatting going on and the newline should probably be handled differently, it would be more clear if this were simpler:

```java
  String errorMessage = response.getStatusMessage() + (response.getContent() != null ? "\n" + new String(...) : "");
```

It would probably make sense to have a private helper method for converting the content string.